scene/main/viewport.cpp

```diff
@@ -1797,104 +1802,155 @@ void Viewport::_gui_input_event(Ref<InputEvent> p_event) {
+						} else {
+							// Test if a SubViewport is available.
+							SubViewportContainer *svc = Object::cast_to<SubViewportContainer>(drag_over_control);
+							if (svc) {
+								if (svc->is_stretch_enabled()) {
+									control_pos /= svc->get_stretch_shrink();
+								}
 
-					if (object_under != ObjectID()) { // Fetch window.
-						Window *w = Object::cast_to<Window>(ObjectDB::get_instance(object_under));
-						if (w) {
-							viewport_under = w;
-							viewport_pos = screen_mouse_pos - w->get_position();
+								for (int i = 0; i < svc->get_child_count(); i++) {
+									SubViewport *c = Object::cast_to<SubViewport>(svc->get_child(i));
+									if (!c || c->is_input_disabled()) {
+										continue;
+									}
+									target_viewport = c;
+									viewport_pos = c->get_final_transform().affine_inverse().xform(control_pos);
+									restart = true;
+									break; // Take only the first available SubViewport into consideration.
+								}
+							}
 						}
 					}
-				}
-			}
 
-			if (viewport_under) {
-				if (viewport_under != this) {
-					Transform2D ai = (viewport_under->get_final_transform().affine_inverse() * viewport_under->_get_input_pre_xform());
-					viewport_pos = ai.xform(viewport_pos);
+					if (restart) {
+						// SubViewport found: Stop checking parents.
+						break;
+					}
+
+					// Update variables for checking parent CanvasItem.
+					viewport_pos = ci->get_transform().xform(viewport_pos);
+					ci = ci->get_parent_item();
 				}
-				// Find control under at position.
-				gui.drag_mouse_over = viewport_under->gui_find_control(viewport_pos);
-				if (gui.drag_mouse_over) {
-					Transform2D localizer = gui.drag_mouse_over->get_global_transform_with_canvas().affine_inverse();
-					gui.drag_mouse_over_pos = localizer.xform(viewport_pos);
 
-					bool can_drop = _gui_drop(gui.drag_mouse_over, gui.drag_mouse_over_pos, true);
+				if (!ci) {
+					// No Drop target was found: End search.
+					break;
+				}
+			}
 
-					if (!can_drop) {
-						ds_cursor_shape = DisplayServer::CURSOR_FORBIDDEN;
-					} else {
-						ds_cursor_shape = DisplayServer::CURSOR_CAN_DROP;
-					}
+			// Send mouse enter/exit notifications to Drop targets.
+			if (previous_drop_target != gui.drag_mouse_over) {
+				Window *start_base_window = get_tree()->get_root()->gui.drag_start_base_window;
+				if (gui.drag_mouse_over && gui.drag_mouse_over->get_viewport()->get_base_window() != start_base_window &&
+						(start_base_window == this || is_ancestor_of(gui.drag_mouse_over->get_viewport()->get_base_window()))) {
+					_gui_call_notification(gui.drag_mouse_over, Control::NOTIFICATION_MOUSE_ENTER);
 				}
+				if (previous_drop_target && previous_drop_target->get_viewport()->get_base_window() != start_base_window &&
+						(start_base_window == this || is_ancestor_of(previous_drop_target->get_viewport()->get_base_window()))) {
+					_gui_call_notification(previous_drop_target, Control::NOTIFICATION_MOUSE_EXIT);
+				}
```

Comment:
This became rather complicated because depending on drag source and target the code is executed in one or more different viewports.
While this current solution works in a large amount of situations, a better solution would be to execute drop-handling only in a single viewport.
For the moment this works for all my testcases, but I intend to revisit this after #58334, since that other PR will change the order in which viewports handle events and so a solution would require adjustment anyway.

Reply:
> [...] a better solution would be to execute drop-handling only in a single viewport

I just verified, that this is not possible, because `SubViewports` can embed `Windows` and that combination makes it necessary to execute this code also in nested `Viewports`.